### PR TITLE
transform from wavelet space to real space with tests

### PR DESCRIPTION
**Pull request by stonecoldhughes**

### CODE REVIEW

File: src/transformations.cpp [hunk 1]
```diff
@@ -75,6 +75,155 @@ fk::vector<P> combine_dimensions(int const degree, element_table const &table,
...
+ * order */
+template<typename P>
+fk::matrix<P>
+recursive_kron(std::vector<fk::matrix<P, mem_type::view>> &kron_matrices,
+               int index)
+{
+  if (index == (static_cast<int>(kron_matrices.size()) - 1))
+    return fk::matrix<P>(kron_matrices.back());
+
+  else
```
Comment (bmcdanie) [inline src/transformations.cpp]: unnecessary else after return in a single line `if` - I'm ok with leaving it in if you prefer to be explicit, matter of preference.

in general, we prefer to use curly braces even for single line conditionals, though I have noticed that some have crept into the code: https://wiki.sei.cmu.edu/confluence/display/c/EXP19-C.+Use+braces+for+the+body+of+an+if%2C+for%2C+or+while+statement.

File: src/transformations_tests.cpp [hunk 2]
```diff
@@ -149,3 +149,76 @@ TEMPLATE_TEST_CASE("forward multi-wavelet transform", "[transformations]",
...
+    int const degree = 3;
+
+    auto pde = make_PDE<TestType>(PDE_opts::continuity_3, level, degree);
+
+    element_table table(make_options({"-l", std::to_string(level)}),
+                        pde->num_dims);
+
+    std::vector<TestType> wave_space;
+
+    int wave_space_size = table.size() * std::pow(degree, pde->num_dims);
```
Comment (bmcdanie) [inline src/transformations_tests.cpp]: `int const`

File: src/transformations_tests.cpp [hunk 3]
```diff
@@ -149,3 +149,76 @@ TEMPLATE_TEST_CASE("forward multi-wavelet transform", "[transformations]",
...
+  };
+
+  SECTION("gen_realspace_transform")
+  {
+    /* Defaults to match gold */
+    int const level = 4;
+
+    int const degree = 3;
+
+    auto pde = make_PDE<TestType>(PDE_opts::continuity_3, level, degree);
```
Comment (bmcdanie) [inline src/transformations_tests.cpp]: `auto const`

File: src/transformations.cpp [hunk 4]
```diff
@@ -75,6 +75,155 @@ fk::vector<P> combine_dimensions(int const degree, element_table const &table,
...
+                      element_table const &table)
+{
+  /* determine the length of the real-space vector */
+  int prod = 1;
+
+  std::vector<dimension<P>> const &dims = pde.get_dimensions();
+
+  /* Assume the degree in the first dimension is equal across all the remaining
+   */
+  int stride = std::pow(pde.get_dimensions()[0].get_degree(), dims.size());
```
[the remark this answers is not in the document]
Comment (stonecoldhughes) [reply src/transformations.cpp]: done

File: src/transformations.cpp [hunk 5]
```diff
@@ -75,6 +75,155 @@ fk::vector<P> combine_dimensions(int const degree, element_table const &table,
...
+  return real_space_transform;
+}
+
+template<typename P>
+fk::vector<P>
+wavelet_to_real_space(PDE<P> const &pde, fk::vector<P> const &wave_space,
+                      element_table const &table)
+{
+  /* determine the length of the real-space vector */
+  int prod = 1;
```
[the remark this answers is not in the document]
Comment (stonecoldhughes) [reply src/transformations.cpp]: done

File: src/transformations.cpp [hunk 6]
```diff
@@ -75,6 +75,155 @@ fk::vector<P> combine_dimensions(int const degree, element_table const &table,
...
+  /* contains a basis matrix for each dimension */
+  std::vector<fk::matrix<P>> real_space_transform;
+
+  real_space_transform.reserve(pde.num_dims);
+
+  for (int i = 0; i < pde.num_dims; i++)
+  {
+    /* get the ith dimension */
+    dimension<P> const &d = pde.get_dimensions()[i];
+
```
[the remark this answers is not in the document]
Comment (stonecoldhughes) [reply src/transformations.cpp]: they gone now

File: src/transformations.cpp [hunk 7]
```diff
@@ -75,6 +75,155 @@ fk::vector<P> combine_dimensions(int const degree, element_table const &table,
...
+    int const deg_freedom_1d = d.get_degree() * n_segments;
+
+    P const normalize = (d.domain_max - d.domain_min) / n_segments;
+
+    fk::matrix<P> dimension_transform(deg_freedom_1d, deg_freedom_1d);
+
+    /* create matrix of Legendre polynomial basis functions evaluated at the
+     * roots */
+    static const auto roots = legendre_weights<P>(d.get_degree(), -1, 1)[0];
```
[the remark this answers is not in the document]
Comment (stonecoldhughes) [reply src/transformations.cpp]: reppin east const

File: src/transformations_tests.cpp (hunk 3 above)
Comment (stonecoldhughes) [reply src/transformations_tests.cpp]: done

File: src/transformations_tests.cpp (hunk 2 above)
Comment (stonecoldhughes) [reply src/transformations_tests.cpp]: done

File: src/transformations.cpp [hunk 8]
```diff
@@ -75,6 +75,155 @@ fk::vector<P> combine_dimensions(int const degree, element_table const &table,
...
+    /* normalized legendre transformation matrix. Column i is legendre
+       polynomial of degree i. element (i, j) is polynomial evaluated at jth
+       root of the highest degree polynomial */
+    fk::matrix<P> const basis = legendre<P>(roots, d.get_degree())[0] *
+                                (static_cast<P>(1.0) / std::sqrt(normalize));
+
+    /* set submatrices of dimension_transform */
+    for (int j = 0; j < n_segments; j++)
+    {
+      int diagonal_pos = d.get_degree() * j;
```
[the remark this answers is not in the document]
Comment (stonecoldhughes) [reply src/transformations.cpp]: done

File: src/transformations_tests.cpp [hunk 9]
```diff
@@ -149,3 +149,76 @@ TEMPLATE_TEST_CASE("forward multi-wavelet transform", "[transformations]",
     relaxed_comparison(gold, test);
   }
 }
+
+TEMPLATE_TEST_CASE("wavelet to realspace transformation", "", double, float)
```
[the remark this answers is not in the document]
Comment (stonecoldhughes) [reply src/transformations_tests.cpp]: done

File: src/transformations_tests.cpp [hunk 10]
```diff
@@ -149,3 +149,181 @@ TEMPLATE_TEST_CASE("forward multi-wavelet transform", "[transformations]",
...
```
Comment (bmcdanie) [inline src/transformations_tests.cpp]: `"function being tested/description", "[component_name]", types..."`

File: src/transformations_tests.cpp [hunk 11]
```diff
@@ -149,3 +149,181 @@ TEMPLATE_TEST_CASE("forward multi-wavelet transform", "[transformations]",
...
+    int const level  = 3;
+    int const degree = 4;
+    auto const pde = make_PDE<TestType>(PDE_opts::continuity_1, level, degree);
+    element_table table(make_options({"-l", std::to_string(level)}),
+                        pde->num_dims);
+    std::vector<TestType> wave_space;
+    int const wave_space_size = table.size() * std::pow(degree, pde->num_dims);
+
+    /* arbitrary function to transform from wavelet space to real space */
+    auto arbitrary_func = [wave_space_size](TestType x) -> TestType {
```
Comment (bmcdanie) [inline src/transformations_tests.cpp]: you're not using `wave_space_size` here, generates a compiler warning

also, add `const`

File: src/transformations_tests.cpp [hunk 12]
```diff
@@ -149,3 +149,181 @@ TEMPLATE_TEST_CASE("forward multi-wavelet transform", "[transformations]",
...
+    fk::vector<TestType> const gold =
+        fk::vector<TestType>(read_vector_from_txt_file(gold_file_name));
+    relaxed_comparison(gold, real_space);
+  }
+
+  /* continuity_2 */
+  SECTION("gen_realspace_transform_2")
+  {
+    /* Defaults to match gold */
+    int const level  = 3;
```
Comment (bmcdanie) [inline src/transformations_tests.cpp]: can we test some of these at different level/degree

File: src/transformations_tests.cpp (hunk 10 above)
Comment (stonecoldhughes) [reply src/transformations_tests.cpp]: done

File: src/transformations_tests.cpp (hunk 11 above)
Comment (stonecoldhughes) [reply src/transformations_tests.cpp]: done

File: src/transformations_tests.cpp (hunk 12 above)
Comment (stonecoldhughes) [reply src/transformations_tests.cpp]: done

**Pull request merged**